File: backend/app/risk.py
from datetime import datetime, timedelta
from typing import List

from app.models import ClassSession, Deadline, Duty, Faculty, RiskFactors


def parse_tm(value: str) -> datetime:
    return datetime.strptime(value, "%H:%M")


def analyze_schedule(sessions: List[ClassSession]) -> dict:
    by_day: dict[str, List[ClassSession]] = {}
    for s in sessions:
        by_day.setdefault(s.day, []).append(s)

    total_hours = 0.0
    consecutive_blocks = 0
    max_consecutive_hours = 0.0
    min_break = float("inf")
    breaks: List[int] = []

    for day_sessions in by_day.values():
        day_sessions.sort(key=lambda s: parse_tm(s.start_time))
        run_hours = 0.0
        prev_end = None
        for s in day_sessions:
            start, end = parse_tm(s.start_time), parse_tm(s.end_time)
            hours = (end - start).total_seconds() / 3600
            total_hours += hours
            if prev_end is not None:
                gap = (start - prev_end).total_seconds() / 60
                breaks.append(gap)
                if gap <= 30:
                    consecutive_blocks += 1
                    run_hours += hours
                else:
                    max_consecutive_hours = max(max_consecutive_hours, run_hours)
                    run_hours = hours
                min_break = min(min_break, gap)
            else:
                run_hours += hours
            prev_end = end
        max_consecutive_hours = max(max_consecutive_hours, run_hours)

    if not breaks:
        min_break = 0.0
    avg_break = sum(breaks) / len(breaks) if breaks else 0.0

    return {
        "total_hours": round(total_hours, 1),
        "consecutive_blocks": consecutive_blocks,
        "max_consecutive_hours": round(max_consecutive_hours, 1),
        "avg_break_minutes": round(avg_break, 1),
        "min_break_minutes": round(min_break, 1),
    }


def _clamp(v: float) -> float:
    return max(0.0, min(100.0, v))


def level_for(score: float) -> str:
    if score >= 70:
        return "Critical"
    if score >= 50:
        return "High"
    if score >= 30:
        return "Moderate"
    return "Low"


def compute_risk(faculty: Faculty, sessions: List[ClassSession],
                 duties: List[Duty], deadlines: List[Deadline]) -> tuple[float, RiskFactors]:
    sessions = [s if isinstance(s, ClassSession) else ClassSession(**s) for s in sessions]
    duties = [d if isinstance(d, Duty) else Duty(**d) for d in duties]
    deadlines = [d if isinstance(d, Deadline) else Deadline(**d) for d in deadlines]
    sch = analyze_schedule(sessions)
    duties_hours = round(sum(d.hours_per_week for d in duties), 1)
    now = datetime.now()
    upcoming = [d for d in deadlines if now <= d.due_date <= now + timedelta(days=14)]
    deadline_pressure = round(
        sum(d.effort_hours / max(1.0, (d.due_date - now).days) for d in upcoming), 2
    )

    total_hours = round(sch["total_hours"] + duties_hours, 1)
    contract = faculty.contract_hours * faculty.fte
    load_ratio = round(total_hours / contract, 2) if contract else 1.0

    # Weighted sub-scores (each 0-100)
    load_score = _clamp((load_ratio - 1.0) * 150)
    schedule_score = _clamp(
        (sch["consecutive_blocks"] * 15)
        + (max(0.0, sch["max_consecutive_hours"] - 3.0) * 12)
        + (max(0.0, 30.0 - sch["min_break_minutes"]) * 1.5)
    )
    duty_score = _clamp((duties_hours * 6) + (len(duties) * 4))
    deadline_score = _clamp(deadline_pressure * 30)

    score = _clamp(
        load_score * 0.30
        + schedule_score * 0.25
        + duty_score * 0.20
        + deadline_score * 0.15
        + (load_score * 0.10)  # stretch bonus for utilization on top
    )

    level = level_for(score)

    factors = RiskFactors(
        total_hours=total_hours,
        contract_hours=contract,
        load_ratio=load_ratio,
        consecutive_blocks=sch["consecutive_blocks"],
        max_consecutive_hours=sch["max_consecutive_hours"],
        avg_break_minutes=sch["avg_break_minutes"],
        min_break_minutes=sch["min_break_minutes"],
        duties_hours=duties_hours,
        duties_count=len(duties),
        deadlines_days=len(upcoming),
        deadline_pressure=deadline_pressure,
    )
    return round(score, 1), factors

File: server/src/services/risk.js
const MINUTES_PER_HOUR = 60;
const DAY_MS = 24 * 60 * 60 * 1000;
const FOURTEEN_DAYS_MS = 14 * DAY_MS;

export function parseTm(value) {
  const [h, m] = String(value).split(":").map(Number);
  return h * 60 + m;
}

export function analyzeSchedule(sessions) {
  const byDay = new Map();
  for (const s of sessions) {
    if (!byDay.has(s.day)) byDay.set(s.day, []);
    byDay.get(s.day).push(s);
  }

  let totalHours = 0;
  let consecutiveBlocks = 0;
  let maxConsecutiveHours = 0;
  let minBreak = Infinity;
  const breaks = [];

  for (const daySessions of byDay.values()) {
    daySessions.sort((a, b) => parseTm(a.start_time) - parseTm(b.start_time));
    let runHours = 0;
    let prevEnd = null;
    for (const s of daySessions) {
      const start = parseTm(s.start_time);
      const end = parseTm(s.end_time);
      const hours = (end - start) / MINUTES_PER_HOUR;
      totalHours += hours;
      if (prevEnd !== null) {
        const gap = (start - prevEnd) / MINUTES_PER_HOUR;
        breaks.push(gap);
        if (gap <= 30) {
          consecutiveBlocks += 1;
          runHours += hours;
        } else {
          maxConsecutiveHours = Math.max(maxConsecutiveHours, runHours);
          runHours = hours;
        }
        minBreak = Math.min(minBreak, gap);
      } else {
        runHours += hours;
      }
      prevEnd = end;
    }
    maxConsecutiveHours = Math.max(maxConsecutiveHours, runHours);
  }

  if (breaks.length === 0) minBreak = 0;
  const avgBreak = breaks.length ? breaks.reduce((a, b) => a + b, 0) / breaks.length : 0;

  return {
    total_hours: round1(totalHours),
    consecutive_blocks: consecutiveBlocks,
    max_consecutive_hours: round1(maxConsecutiveHours),
    avg_break_minutes: round1(avgBreak),
    min_break_minutes: round1(minBreak),
  };
}

function round1(v) {
  return Math.round(v * 10) / 10;
}

function clamp(v) {
  return Math.max(0, Math.min(100, v));
}

export function levelFor(score) {
  if (score >= 70) return "Critical";
  if (score >= 50) return "High";
  if (score >= 30) return "Moderate";
  return "Low";
}

export function computeRisk(faculty, sessions, duties, deadlines) {
  const sch = analyzeSchedule(sessions);
  const dutiesHours = round1(duties.reduce((sum, d) => sum + (Number(d.hours_per_week) || 0), 0));

  const now = Date.now();
  const upcoming = deadlines.filter((d) => {
    const due = new Date(d.due_date).getTime();
    return now <= due && due <= now + FOURTEEN_DAYS_MS;
  });
  const deadlinePressure = round2(
    upcoming.reduce((sum, d) => {
      const days = Math.max(1, Math.floor((new Date(d.due_date).getTime() - now) / DAY_MS));
      return sum + (Number(d.effort_hours) || 0) / days;
    }, 0)
  );

  const totalHours = round1(sch.total_hours + dutiesHours);
  const contract = Number(faculty.contract_hours || 0) * Number(faculty.fte || 0);
  const loadRatio = contract ? round2(totalHours / contract) : 1.0;

  const loadScore = clamp((loadRatio - 1.0) * 150);
  const scheduleScore = clamp(
    sch.consecutive_blocks * 15 +
      Math.max(0, sch.max_consecutive_hours - 3.0) * 12 +
      Math.max(0, 30.0 - sch.min_break_minutes) * 1.5
  );
  const dutyScore = clamp(dutiesHours * 6 + duties.length * 4);
  const deadlineScore = clamp(deadlinePressure * 30);

  const score = clamp(
    loadScore * 0.3 + scheduleScore * 0.25 + dutyScore * 0.2 + deadlineScore * 0.15 + loadScore * 0.1
  );

  const factors = {
    total_hours: totalHours,
    contract_hours: round1(contract),
    load_ratio: loadRatio,
    consecutive_blocks: sch.consecutive_blocks,
    max_consecutive_hours: sch.max_consecutive_hours,
    avg_break_minutes: sch.avg_break_minutes,
    min_break_minutes: sch.min_break_minutes,
    duties_hours: dutiesHours,
    duties_count: duties.length,
    deadlines_days: upcoming.length,
    deadline_pressure: deadlinePressure,
  };

  return { score: round1(score), level: levelFor(score), factors };
}

function round2(v) {
  return Math.round(v * 100) / 100;
}